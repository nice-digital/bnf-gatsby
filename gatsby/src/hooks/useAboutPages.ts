import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface AboutPages {
	allBnfAboutSection: {
		aboutSectionPages: SlugAndTitle[];
	};
	allBnfCautionaryAndAdvisoryGuidance: {
		labelsGuidancePages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/about/${slug}/`,
});

/**
 * Gets the lists of pages for the about section by combining:
 * - about sections
 * - cautionary/advisory labels
 * - cautionary/advisory guidance page(s)
 * from the feed.
 *
 * @returns The consolidated list of about pages
 */
export const useAboutPages = (): MenuPageLink[] => {
	const { allBnfAboutSection, allBnfCautionaryAndAdvisoryGuidance } =
		useStaticQuery<AboutPages>(
			graphql`
				query AboutPages {
					allBnfAboutSection(sort: { order: ASC }) {
						aboutSectionPages: nodes {
							slug
							title
						}
					}
					allBnfCautionaryAndAdvisoryGuidance {
						labelsGuidancePages: nodes {
							title
							slug
						}
					}
				}
			`
		);

	return [
		...allBnfAboutSection.aboutSectionPages.map(slugToHref),
		{ title: "Labels", href: "/about/labels/" },
		...allBnfCautionaryAndAdvisoryGuidance.labelsGuidancePages.map(slugToHref),
	];
};
