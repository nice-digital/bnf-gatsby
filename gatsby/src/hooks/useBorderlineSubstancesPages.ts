import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
