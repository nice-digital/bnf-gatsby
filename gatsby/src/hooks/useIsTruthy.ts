import { useCallback } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useIsTruthy = () =>
	useCallback(
		<TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe => !!maybeT,
		[]
	);
