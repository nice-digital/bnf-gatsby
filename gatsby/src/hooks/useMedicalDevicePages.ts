import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicalDevicesQueryResult {
	allBnfMedicalDevice: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medical-devices/${slug}/`,
});

/**
 * Gets the lists of pages for the medical devices section
 *
 * @returns The consolidated list of medical devices pages
 */
export const useMedicalDevicePages = (): MenuPageLink[] => {
	const { allBnfMedicalDevice } = useStaticQuery<MedicalDevicesQueryResult>(
		graphql`
			{
				allBnfMedicalDevice(sort: { fields: title, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfMedicalDevice.pages.map(slugToHref);
};
