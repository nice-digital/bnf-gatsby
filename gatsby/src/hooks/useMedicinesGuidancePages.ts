import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicinesGuidanceQueryResult {
	allBnfGuidance: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medicines-guidance/${slug}/`,
});

/**
 * Gets the lists of pages for the medicines guidance section
 *
 * @returns The consolidated list of medicines guidance pages
 */
export const useMedicinesGuidancePages = (): MenuPageLink[] => {
	const { allBnfGuidance } = useStaticQuery<MedicinesGuidanceQueryResult>(
		graphql`
			{
				allBnfGuidance(sort: { fields: order, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfGuidance.pages.map(slugToHref);
};
