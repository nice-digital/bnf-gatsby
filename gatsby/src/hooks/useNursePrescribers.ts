import { useStaticQuery, graphql } from "gatsby";

import { SlugAndTitle, type MenuPageLink } from "@/utils";

export interface NursePrescribers {
	allBnfNursePrescribersFormularyIntroduction: {
		introduction: SlugAndTitle[];
	};
	allBnfNursePrescribersFormularyTreatmentSummary: {
		summaries: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/nurse-prescribers-formulary/${slug}/`,
});

/**
 * Gets the lists of pages for the nurse prescribers formulary by combining:
 * - treatment summaries
 * - introduction
 * from the feed.
 *
 * @returns both the consolidated list for the details page menu, and two separate lists for the index page.
 */
export const useNursePrescribers = (): {
	menuList: MenuPageLink[];
	aboutList: MenuPageLink[];
	treatmentSummariesList: MenuPageLink[];
} => {
	const {
		allBnfNursePrescribersFormularyIntroduction,
		allBnfNursePrescribersFormularyTreatmentSummary,
	} = useStaticQuery<NursePrescribers>(
		graphql`
			query NursePrescribers {
				allBnfNursePrescribersFormularyIntroduction {
					introduction: nodes {
						slug
						title
					}
				}
				allBnfNursePrescribersFormularyTreatmentSummary {
					summaries: nodes {
						slug
						title
					}
				}
			}
		`
	);

	const menuList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries.map(
			slugToHref
		),
		...allBnfNursePrescribersFormularyIntroduction.introduction.map(slugToHref),
	];

	const aboutList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries
			.filter((summary) => summary.slug == "general-guidance")
			.map(slugToHref),
		...allBnfNursePrescribersFormularyIntroduction.introduction.map(slugToHref),
	];

	const treatmentSummariesList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries.map(
			slugToHref
		),
	];

	return { menuList, aboutList, treatmentSummariesList };
};
