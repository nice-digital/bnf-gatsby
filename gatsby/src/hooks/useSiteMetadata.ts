import { useStaticQuery, graphql } from "gatsby";

interface SiteMetaData {
	/** Whether the current site is BNF or BNFC */
	isBNF: boolean;
	/** Absolute URL to the site homepage, without trailing slash */
	siteUrl: string;
	/** Short site title e.g. "BNF" or "BNFC" */
	siteTitleShort: string;
	/** Full site title e.g. "British National Formulary" or "British National Formulary for Children" */
	siteTitleLong: string;
}

interface Site {
	site: {
		siteMetadata: SiteMetaData;
	};
}

/**
 * Gets the site metadata, as defined in gatsby-config
 *
 * @returns The site metadata
 */
export const useSiteMetadata = (): SiteMetaData => {
	const { site } = useStaticQuery<Site>(
		graphql`
			query SiteMetaData {
				site {
					siteMetadata {
						isBNF
						siteUrl
						siteTitleShort
						siteTitleLong
					}
				}
			}
		`
	);
	return site.siteMetadata;
};
