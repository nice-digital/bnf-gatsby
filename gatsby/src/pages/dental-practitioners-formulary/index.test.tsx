import { render, screen, waitFor } from "@testing-library/react";

import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import DentalPractitionersFormularyPage, {
	type DentalPractitionersFormularyPageProps,
	query,
} from "./";

const pageProps: DentalPractitionersFormularyPageProps = {
	data: {
		bnfDentalPractitionersFormulary: {
			title: "Dental Practitioners’ Formulary",
			sections: [
				{
					content:
						"<p>The following list has been approved by the appropriate Secretaries of State</p>",
					slug: "list-of-dental-preparations",
					title: "List of Dental Preparations",
				},
				{
					content:
						"<p>Preparations on the List of Dental Preparations which are specified as DPF are described as follows in the DPF.</p>",
					slug: "details-of-dpf-preparations",
					title: "Details of DPF preparations",
				},
			],
		},
	},
};

describe("DentalPractitionersFormularyPageProps", () => {
	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	it("should match snapshot for page contents", () => {
		render(<DentalPractitionersFormularyPage {...pageProps} />);

		expect(screen.getByRole("main")).toMatchSnapshot();
	});

	it("should set page title", async () => {
		render(<DentalPractitionersFormularyPage {...pageProps} />);

		await waitFor(() => {
			expect(document.title).toStartWith("Dental Practitioners’ Formulary |");
		});
	});

	it("should render meta description for BNF", async () => {
		render(<DentalPractitionersFormularyPage {...pageProps} />);

		await waitFor(() => {
			expect(
				document
					// eslint-disable-next-line testing-library/no-node-access
					.querySelector("meta[name='description']")
			).toHaveAttribute(
				"content",
				"Browse the Dental Practitioners' Formulary (DPF) - the list of approved preparations for prescribing by dentists in the BNF."
			);
		});
	});

	it("should render meta description for BNFC", async () => {
		(useSiteMetadata as jest.Mock).mockImplementationOnce(() => ({
			siteTitleShort: "BNFC",
		}));

		render(<DentalPractitionersFormularyPage {...pageProps} />);

		await waitFor(() => {
			expect(
				document
					// eslint-disable-next-line testing-library/no-node-access
					.querySelector("meta[name='description']")
			).toHaveAttribute(
				"content",
				"Browse the Dental Practitioners' Formulary (DPF) - the list of approved preparations for prescribing by dentists in the BNFC."
			);
		});
	});
});
