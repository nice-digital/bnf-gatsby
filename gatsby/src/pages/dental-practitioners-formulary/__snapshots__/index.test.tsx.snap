// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DentalPractitionersFormularyPageProps should match snapshot for graphql query 1`] = `
"
	{
		bnfDentalPractitionersFormulary {
			title
			sections {
				...RecordSection
			}
		}
	}
"
`;

exports[`DentalPractitionersFormularyPageProps should match snapshot for page contents 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="https://www.nice.org.uk/"
        >
          NICE
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/"
        >
          BNF
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Dental Practitioners’ Formulary
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Dental Practitioners’ Formulary"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
    id="content-start"
  >
    <h1
      class="page-header__heading"
    >
      <span>
        Dental Practitioners’ Formulary
      </span>
    </h1>
  </div>
  <div
    class="grid grid--loose"
    data-testid="body"
  >
    <div
      data-g="12 md:12 lg:12"
    >
      <div
        class="grid grid--loose"
      >
        <div
          data-g="12 md:8 lg:9"
        >
          <nav
            aria-labelledby="navigate-to-section"
            class="nav"
          >
            <h2
              class="heading"
              id="navigate-to-section"
            >
              <span>
                Navigate to section
              </span>
            </h2>
            <ol
              aria-label="Jump links to sections on this page"
              class="linkList"
            >
              <li>
                <a
                  href="#list-of-dental-preparations"
                >
                  List of Dental Preparations
                </a>
              </li>
              <li>
                <a
                  href="#details-of-dpf-preparations"
                >
                  Details of DPF preparations
                </a>
              </li>
            </ol>
          </nav>
        </div>
        <div
          class="body"
          data-g="12 md:8 lg:9"
        >
          <section
            aria-labelledby="list-of-dental-preparations"
          >
            <h2
              class="firstHeading"
              id="list-of-dental-preparations"
            >
              List of Dental Preparations
            </h2>
            <div>
              <p>
                The following list has been approved by the appropriate Secretaries of State
              </p>
            </div>
          </section>
          <section
            aria-labelledby="details-of-dpf-preparations"
          >
            <h2
              id="details-of-dpf-preparations"
            >
              Details of DPF preparations
            </h2>
            <div>
              <p>
                Preparations on the List of Dental Preparations which are specified as DPF are described as follows in the DPF.
              </p>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;
